core/src/main/java/org/apache/druid/java/util/common/HumanReadableBytes.java

```diff
@@ -128,71 +135,76 @@ public static long parse(String number, long nullValue)
-    char unit = number.charAt(index--);
+    String number = StringUtils.toLowerCase(rawNumber);
+    if (number.charAt(0) == '-') {
+      throw new IAE("Invalid format of number: %s. Negative value is not allowed.", rawNumber);
+    }
+
+    int lastDigitIndex = number.length() - 1;
+    boolean isBinaryByte = false;
+    char unit = number.charAt(lastDigitIndex--);
     if (unit == 'b') {
-      if (index < 2) {
-        throw new IAE("invalid format of number[%s]", number);
+      //unit ends with 'b' must be format of KiB/MiB/GiB/TiB/PiB, so at least 3 extra characters are required
+      if (lastDigitIndex < 2) {
+        throw new IAE("Invalid format of number: %s", rawNumber);
       }
-      if (number.charAt(index--) != 'i') {
-        throw new IAE("invalid format of number[%s]", number);
+      if (number.charAt(lastDigitIndex--) != 'i') {
+        throw new IAE("Invalid format of number: %s", rawNumber);
       }
 
-      unit = number.charAt(index--);
-      isBinary = true;
+      unit = number.charAt(lastDigitIndex--);
+      isBinaryByte = true;
     }
+
     long base = 1;
     switch (unit) {
       case 'k':
-        base = isBinary ? 1024 : 1_000;
+        base = isBinaryByte ? 1024 : 1_000;
         break;
 
       case 'm':
-        base = isBinary ? 1024 * 1024 : 1_000_000;
+        base = isBinaryByte ? 1024 * 1024 : 1_000_000;
         break;
 
       case 'g':
-        base = isBinary ? 1024 * 1024 * 1024 : 1_000_000_000;
+        base = isBinaryByte ? 1024 * 1024 * 1024 : 1_000_000_000;
         break;
 
       case 't':
-        base = isBinary ? 1024 * 1024 * 1024 * 1024L : 1_000_000_000_000L;
+        base = isBinaryByte ? 1024 * 1024 * 1024 * 1024L : 1_000_000_000_000L;
         break;
 
       case 'p':
-        base = isBinary ? 1024L * 1024 * 1024 * 1024 * 1024 : 1_000_000_000_000_000L;
+        base = isBinaryByte ? 1024L * 1024 * 1024 * 1024 * 1024 : 1_000_000_000_000_000L;
         break;
 
       default:
+        lastDigitIndex++;
         if (!Character.isDigit(unit)) {
-          throw new IAE("invalid character in number[%s]", number);
+          throw new IAE("Invalid format of number: %s", rawNumber);
         }
         break;
     }
 
     try {
-      if (base > 1 && index >= 0) {
-        long value = Long.parseLong(number.substring(0, index + 1)) * base;
-        if (value < base) {
-          throw new IAE("number [%s] overflow", number);
-        }
-        return value;
-      } else {
-        return Long.parseLong(number);
+      long value = Long.parseLong(number.substring(0, lastDigitIndex + 1)) * base;
+      if (base > 1 && value < base) {
+        //for base == 1, overflow has been checked in parseLong
+        throw new IAE("Number overflow: %s", rawNumber);
       }
+      return value;
     }
     catch (NumberFormatException e) {
-      throw new IAE("invalid format of number[%s]", number);
+      throw new IAE("Invalid format of number: %s", rawNumber);
```

Comment:
NumberFormatException throws when 1) there're invalid characters ; 2) number overflows. It's unknown for us which case it is when it's thrown.

Reply:
I meant the error message should be understandable for Druid operators. The current message is confusing since it doesn't mention another possibility for out of range. Perhaps it's better to be `Invalid number format or out of range of long: %s`.